src/rest/connection.cpp

```diff
@@ -0,0 +1,373 @@
+        break;
+
+    case OTBR_REST_CONNECTION_INTERNALERROR:
+
+        mResource->ErrorHandler(mRequest, mResponse, 500);
+        error = Write();
+        break;
+
+    default:
+        break;
+    }
+    if (error == OTBR_ERROR_REST)
+    {
+        Disconnect();
+    }
+
+    return error;
+}
+
+otbrError Connection::Handle()
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    shutdown(mFd, SHUT_RD);
+
+    mResource->Handle(mRequest, mResponse);
+
+    if (mResponse.NeedCallback())
+    {
+        // Set its state as Waitforcallback and refresh its timer
+        mState     = OTBR_REST_CONNECTION_CALLBACKWAIT;
+        mStartTime = steady_clock::now();
+    }
+    else
+    {
+        // Normal Write back process.
+        error = Write();
+    }
+
+    return error;
+}
+
+otbrError Connection::ProcessWaitCallback()
+{
+    otbrError error    = OTBR_ERROR_NONE;
+    auto      duration = duration_cast<microseconds>(steady_clock::now() - mStartTime).count();
+
+    if (duration >= kCallbackTimeout)
+    {
+        // Handle Callback, then Write back response
+        mResource->HandleCallback(mRequest, mResponse);
+        error = Write();
+    }
+
+    return error;
+}
+
+otbrError Connection::ProcessWaitWrite(fd_set &aWriteFdSet)
+{
+    otbrError error    = OTBR_ERROR_NONE;
+    auto      duration = duration_cast<microseconds>(steady_clock::now() - mStartTime).count();
+
+    if (duration <= kWriteTimeout)
+    {
+        if (FD_ISSET(mFd, &aWriteFdSet))
+        {
+            // WriteFdSet is set, then try to write again
+            error = Write();
+        }
+    }
+    else
+    { // Pass write timeout, just close the error
+        Disconnect();
+    }
+    return error;
+}
+
+otbrError Connection::Process(fd_set &aReadFdSet, fd_set &aWriteFdSet)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    switch (mState)
+    {
+    // Initial state, directly read for the first time.
+    case OTBR_REST_CONNECTION_INIT:
+    {
+        error = ProcessWaitRead(aReadFdSet);
+        break;
+    }
+    case OTBR_REST_CONNECTION_READWAIT:
+    {
+        error = ProcessWaitRead(aReadFdSet);
+        break;
+    }
+    case OTBR_REST_CONNECTION_CALLBACKWAIT:
+    {
+        //  Wait for Callback process.
+        error = ProcessWaitCallback();
+        break;
+    }
+    case OTBR_REST_CONNECTION_WRITEWAIT:
+    {
+        error = ProcessWaitWrite(aWriteFdSet);
+        break;
+    }
+    default:
+        break;
+    }
+
+    return error;
+}
+
+otbrError Connection::Write()
+{
+    otbrError   error = OTBR_ERROR_NONE;
+    std::string errorCode;
+    int32_t     sendLength;
+    int         err;
+
+    if (mState != OTBR_REST_CONNECTION_WRITEWAIT)
+    {
+        // Change its state when try write for the first time.
+        mState        = OTBR_REST_CONNECTION_WRITEWAIT;
+        mStartTime    = steady_clock::now();
+        mWriteContent = mResponse.Serialize();
+    }
+
+    if (mWriteContent.size() == 0)
+    {
+        Disconnect();
```

Comment:
`Disconnect` will firstly close the socket file descriptor then label it as `complete` so server will clear it in next loop. Seems this function is enough. 
but I found that there is no possibility that `mWriteContent.size()`  is 0, because a response always contain some default header information, so I may well just delete this check.